fpga_interchange/testarch_generators/device_resources_builder.py

```diff
@@ -0,0 +1,565 @@
+#!/usr/bin/env python3
+from enum import Enum
+
+# =============================================================================
+
+
+class Direction(Enum):
+    Input = 0
+    Output = 1
+    Inout = 2
+
+class BelCategory(Enum):
+    LOGIC = 0
+    ROUTING = 1
+    SITE_PORT = 2
+
+class ConstantType(Enum):
+    NO_PREFERENCE = 0
+    GND = 1
+    VCC = 2
+
+class BelPin():
+
+    def __init__(self, name, direction):
+        self.name = name
+        self.direction = direction
+
+    def __repr__(self):
+        return "BelPin(\"{}\", {})".format(self.name, self.direction.name)
+
+class Bel():
+
+    def __init__(self, name, type, category):
+        self.name = name
+        self.type = type
+        self.category = category
+        self.is_inverting = False # TODO: FIXME:
+
+        self.pins = {}      # dict(name) -> BelPin
+
+    def add_pin(self, name, direction):
+        """
+        Adds a pin to the BEL
+        """
+        assert name not in self.pins, name
+        self.pins[name] = BelPin(name, direction)
+
+        return self.pins[name]
+
+    def __repr__(self):
+        return "Bel(\"{}\", \"{}\", {})".format(
+            self.name, self.type, self.category.name)
+
+class SitePin():
+
+    def __init__(self, name, direction, bel_name):
+        self.name = name
+        self.direction = direction
+        self.bel_name = bel_name
+
+    def __repr__(self):
+        return "SitePin(\"{}\", {}, bel=\"{}\")".format(
+            self.name, self.direction.name, self.bel_name)
+
+class SiteWire():
+
+    def __init__(self, name):
+        self.name = name
+        self.bel_pins = set()
+
+    def connect_to_bel_pin(self, bel_name, pin_name):
+        """
+        Connects the wire to the given BEL and its pin. Does not check if the
+        given names are legal!
+        """
+        bel_pin = (bel_name, pin_name)
+        assert bel_pin not in self.bel_pins, bel_pin
+        self.bel_pins.add(bel_pin)
+
+class SitePip():
+
+    def __init__(self, src_bel_pin, dst_bel_pin):
+        self.src_bel_pin = src_bel_pin
+        self.dst_bel_pin = dst_bel_pin
+
+    def __repr__(self):
+        return "SitePip({}, {})".format(self.src_bel_pin, self.dst_bel_pin)
+
+class SiteType():
+
+    def __init__(self, name):
+        self.name = name
+        self.pins = {}      # dict(name) -> SitePin
+        self.bels = {}      # dict(name) -> Bel
+        self.pips = set()
+        self.wires = {}     # dict(name) -> SiteWire
+
+#        self.alt_site_types = []
+
+    def add_pin(self, name, direction):
+        """
+        Adds a pin to the site type along with its corresponding BEL
+        """
+
+        # Add BEL
+        bel_name = name
+        bel = Bel(bel_name, "pin", BelCategory.SITE_PORT)
+        assert bel.name not in self.bels, bel.name
+        self.bels[bel.name] = bel
+
+        # Add BEL pin
+        bel.add_pin(name, direction) # FIXME: Opposite direction ?!?!?!
```

Comment:
Yeah - an input to the site from the tile is an output from the site port bel pin within the tile (think of it a bit like a 'buffer' that brings the signal into the tile)

Reply:
Allright, thanks.